kolibri/core/assets/src/api_resource.js

```diff
@@ -0,0 +1,68 @@
+const logging = require('loglevel');
+const rest = require('rest');
+const mime = require('rest/interceptor/mime');
+
+const Kolibri = require('kolibri');
+
+const client = rest.wrap(mime);
+
+class Model extends Object {
+  constructor(data, { resourceName, idKey, params } = { idKey: 'id', params: {} }) {
+    super(data);
+    this._resourceName = resourceName;
+    if (typeof this._resourceName !== 'string') {
+      throw new TypeError('resourceName must be defined');
+    }
+    this._idKey = idKey;
+    this._params = params;
+  }
+  fetch() {
+    return new Promise((resolve, reject) => {
+      client({ path: this._url(), params: this._params }).then((response) => {
+        Object.assign(this, response.entity);
+        resolve(this);
+      }, (response) => {
+        logging.error('An error occurred', response);
+        reject(response);
+      });
+    });
+  }
+  _url() {
+    return Kolibri.urls[`${this._resourceName}_detail`](this[this._idKey]);
```

Comment:
A couple additional comments here -

1. It seems brittle to dynamically generate lookup keys for objects because it relies on naming conventions. It also makes it hard to search the code base for the the keys and attributes. It looks like it's using `..._list` and `..._details` as a way to dynamically look up the names of Python functions? I may not fully understand how Django's named URL patterns are supposed to be used, but it seems odd to have them "parseable" because it creates an implicit, undocumented API. Dynamically generating URL reverse-lookup names doesn't seem different from dynamically generating the URLs themselves, which are at least explicitly documented.

2. I'm having trouble finding where `Kolibri.urls` is defined, and what its keys and its values are. I think `require('kolibri')` returns the `kolibriGlobal` which is an instance of `core_app_constructor`, but urls must be added somewhere else that I can't find.